### PR TITLE
Emit package events

### LINKED ISSUE #9253: Emit events w/ dependency information
Its useful to be able to understand various information about the set of libraries an application uses, such as the library name, version, checksum, and more. This information can be used to understand the context in which an application is running, or to detect usage of vulnerable dependencies.

In a java application, its possible to inspect every loaded class using javaagent instrumentation. From a class, you can resolve the CodeSource, and the URL of the JAR. Given the URL of a JAR, you can do things like inspect its manifest, find / inspect the POM, compute checksums, and more. From these data sources, you can compute useful metadata about the JAR, including:

- The JAR Manifest “Implementation-Title”, e.g. “Apache Log4j SLF4J 2.0 Binding”
- The JAR Manifest “Implementation-Vendor”, e.g. “The Apache Software Foundation”
- The POM group id, artifact id, version, e.g.: “org.apache.logging.log4j:log4j-slf4j2-impl:2.20.0”
- The JAR file name, e.g. “log4j-slf4j2-impl-2.20.0.jar”
- The JAR SHA-1 checksum, e.g. “155c8b9bbdac91d8461d9a403a646e6bd0d365d8”
- Useful for verifying JAR content as published on mvn, e.g. [org.apache.logging.log4j:log4j-slf4j2-impl:2.20.0.jar.sha1](https://repo1.maven.org/maven2/org/apache/logging/log4j/log4j-slf4j2-impl/2.20.0/log4j-slf4j2-impl-2.20.0.jar.sha1)

I propose we add a javaagent feature to collect this data and emit it as [OpenTelemetry Events](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md), with details as follows:

- Use existing ECS conventions where applicable.
  - [Package Field conventions](https://www.elastic.co/guide/en/ecs/current/ecs-package.html) mapping:
    - package.name: “org.apache.logging.log4j:log4j-slf4j2-impl”
    - package.version: 2.20.0
    - package.type: jar
    - package.description: Apache Log4j SLF4J 2.0 Binding by The Apache Software Foundation
    - package.checksum: 155c8b9bbdac91d8461d9a403a646e6bd0d365d8
    - package.path: log4j-slf4j2-impl-2.20.0.jar
  - ECS [event.category ](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html)and [event.type](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-type.html):
    - event.domain: [package](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html#ecs-event-category-package)
    - event.name: [info](https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-type.html#ecs-event-type-info)
- Rate limit the jar processing.
  - Processing of JARs requires processing the content InputStream, which can use significant CPU / disk IO if unbounded. We should rate limit the process and add a reasonable default limit (e.g. 10 JARs / second).
- Periodically resend the events.
  - Some backends will have data retention limits that would make these dependency events eventually inaccessible for long running JARs. To accommodate this, we should allow this capability to configure resending the data on some sort of interval (e.g. every 24 hours).
- Allow this capability to be turned off.
  - Like other instrumentation, we should allow this to be turned off, and maybe keep it off by default while we work out the details.

I have a prototype otel java agent extension which produces this telemetry. I still need to add some additional error handling for edge cases, but when things go right, it emits log records as follows (OTLP JSON encoding):
```
{
	"timeUnixNano": "1692389915691751000",
	"observedTimeUnixNano": "1692389915691796000",
	"body": {
		"stringValue": ""
	},
	"attributes": [{
		"key": "event.domain",
		"value": {
			"stringValue": "package"
		}
	}, {
		"key": "event.name",
		"value": {
			"stringValue": "info"
		}
	}, {
		"key": "package.checksum",
		"value": {
			"stringValue": "30d16ec2aef6d8094c5e2dce1d95034ca8b6cb42"
		}
	}, {
		"key": "package.description",
		"value": {
			"stringValue": "Jackson datatype: JSR310 by FasterXML"
		}
	}, {
		"key": "package.name",
		"value": {
			"stringValue": "com.fasterxml.jackson.datatype:jackson-datatype-jsr310"
		}
	}, {
		"key": "package.path",
		"value": {
			"stringValue": "jackson-datatype-jsr310-2.15.2.jar"
		}
	}, {
		"key": "package.type",
		"value": {
			"stringValue": "jar"
		}
	}, {
		"key": "package.version",
		"value": {
			"stringValue": "2.15.2"
		}
	}]
}
```

Looking forward to hearing what people think. Happy to contribute this feature we folks agree.

### PR DESCRIPTION
**Pull request by jack-berg**
Resolves #9253.

### CODE REVIEW

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/test/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    List<LogRecordData> logRecordData = testing.logRecords();
+    List<LogRecordData> events =
+        logRecordData.stream()
+            .filter(
+                record -> {
+                  Attributes attributes = record.getAttributes();
+                  return "jvm".equals(attributes.get(AttributeKey.stringKey("event.domain")))
+                      && "info".equals(attributes.get(AttributeKey.stringKey("event.name")));
+                })
+            .peek(logRecord -> System.out.println(logRecord.getAttributes()))
```
Comment (jack-berg) [inline instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/test/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerTest.java:38]: This rather crude test is really just meant to show where I'm at with this. The following content is printed to the console when you run it:
```
{event.domain="jvm", event.name="info", package.checksum="9a259a1df0841f802b48dee8cfab0fb526bd9e95", package.description="Gradle", package.path="gradle-messaging-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="a96d6978cb02654c4d45b14babcd7777566fdd89", package.description="Gradle", package.path="gradle-base-services-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="950782352b1e7b490948b60f6e28435748aa4e8c", package.description="Gradle", package.path="gradle-logging-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="15cf5acff6d5c3e6d8d9389dfaa104696d5f93cd", package.description="Gradle", package.path="gradle-process-services-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="b5a4b6d16ab13e34a88fae84c35cd5d68cac922c", package.description="slf4j-api", package.name="org.slf4j:slf4j-api", package.path="slf4j-api-1.7.30.jar", package.type="jar", package.version="1.7.30"}
{event.domain="jvm", event.name="info", package.checksum="1524e41b99b14b6e79312e67d5ca97fd173cde10", package.description="Gradle", package.path="gradle-logging-api-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="cbe87e9ddf4da3b8fb0619e1c2cdfb5e73c18b46", package.description="Gradle", package.path="gradle-enterprise-logging-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="e200977bcea40414ce41e3fbf22c0a5c5ff45d05", package.description="Gradle", package.path="gradle-native-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="18751f6367987dcb341e0293885a7d78509c27ab", package.description="Gradle", package.path="gradle-build-operations-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="8d465dda8d734dee9820e8cf83f1465eaaab4f99", package.description="Gradle", package.path="gradle-enterprise-workers-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="cf4a57cfcee7fb3e6eb98091f7ce1beb0de54347", package.description="Gradle", package.path="gradle-testing-base-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="77d5d61f0994ba87f92dc73b0f0157cef9503b18", package.description="Gradle", package.path="gradle-testing-jvm-infrastructure-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="66a92110e0267b4af39bbe4f68033ffbb1273c17", package.path="native-platform-0.22-milestone-24.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="78fd0713dc18f5c6af1bebb7f8ced189c9fca78d", package.description="Gradle", package.path="gradle-files-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="48123c815efde7fb26ad9db44a4179505fbc27ce", package.description="Gradle", package.path="gradle-file-temp-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="d58bebff8cbf70ff52b59208586095f467656c30", package.name="org.slf4j:jul-to-slf4j", package.path="jul-to-slf4j-1.7.30.jar", package.type="jar", package.version="1.7.30"}
{event.domain="jvm", event.name="info", package.checksum="1b2514d2e8a13f4c795939528bb30ef48cd99837", package.description="Gradle", package.path="gradle-testing-junit-platform-8.3.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="0c6b206e80cfd97e66a1364003724491c757b92f", package.path="kryo-2.24.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="0ce1edb914c94ebc388f086c6827e8bdeec71ac2", package.description="Commons Lang by The Apache Software Foundation", package.name="commons-lang:commons-lang", package.path="commons-lang-2.6.jar", package.type="jar", package.version="2.6"}
{event.domain="jvm", event.name="info", package.checksum="89a1922534ed102be1fb2a8c0b2c6151297a12bf", package.description="junit-platform-launcher by junit.org", package.path="junit-platform-launcher-1.10.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="276c4edcf64fabb5a139fa7b4f99330d7a93b804", package.description="junit-platform-engine by junit.org", package.path="junit-platform-engine-1.10.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="d533ff2c286eaf963566f92baf5f8a06628d2609", package.description="junit-platform-commons by junit.org", package.path="junit-platform-commons-1.10.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="90587932d718fc51a48112d33045a18476c542ad", package.description="junit-jupiter-engine by junit.org", package.path="junit-jupiter-engine-5.10.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="af4e0a3cb6901aa53d6401003fc10638014b39b1", package.description="junit-vintage-engine by junit.org", package.path="junit-vintage-engine-5.10.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="653d1b0abf992ac0b3b7262991a8b73bce648497", package.description="spock-core by spockframework.org", package.path="spock-core-2.4-M1-groovy-4.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="8b528fa0a49297582932bdd9096c84dedf71cf25", package.path="junit-pioneer-1.9.1.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="8ac9e16d933b6fb43bc7f576336b8f4d7eb5ba12", package.description="JUnit by JUnit", package.path="junit-4.13.2.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="2fe4ba3d31d5067878e468c96aa039005a9134d3", package.description="junit-jupiter-api by junit.org", package.path="junit-jupiter-api-5.10.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="7c85ec12689813fdcfb66ab9b7533184e169ce1d", package.description="testing-common by OpenTelemetry", package.path="opentelemetry-testing-common-1.30.0-alpha-SNAPSHOT.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="68151d4465c12183db8edc4f9d8f0a878bada16b", package.description="logs", package.path="opentelemetry-sdk-logs-1.29.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="4d512d413b426037b7027fd39426542748ed4581", package.description="Groovy: a powerful, multi-faceted language for the JVM by The Apache Software Foundation", package.path="groovy-4.0.14.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="b7414c6d6a05b9f65b964689abd19a2c236783a6", package.description="spock-junit4 by spockframework.org", package.path="spock-junit4-2.4-M1-groovy-4.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="1820c0968dba3a11a1b30669bb1f01978a91dedc", package.description="hamcrest by hamcrest.org", package.path="hamcrest-2.2.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="2c39784846001a9cffd6c6b89c78de62c0d80fb8", package.name="org.awaitility:awaitility", package.path="awaitility-4.2.0.jar", package.type="jar", package.version="4.2.0"}
{event.domain="jvm", event.name="info", package.checksum="41eb7184ea9d556f23e18b5cb99cad1f8581fc00", package.description="slf4j-api", package.name="org.slf4j:slf4j-api", package.path="slf4j-api-2.0.7.jar", package.type="jar", package.version="2.0.7"}
{event.domain="jvm", event.name="info", package.checksum="e003985c3b9a8865c6bbc938864235cdeea3f516", package.description="Logback Classic Module by QOS.ch", package.name="ch.qos.logback:logback-classic", package.path="logback-classic-1.3.8.jar", package.type="jar", package.version="1.3.8"}
{event.domain="jvm", event.name="info", package.checksum="27d73ef908f832790b9eda936dd6476cc22e47c5", package.description="Logback Core Module by QOS.ch", package.name="ch.qos.logback:logback-core", package.path="logback-core-1.3.8.jar", package.type="jar", package.version="1.3.8"}
{event.domain="jvm", event.name="info", package.checksum="45010687a1181dc886fd12403e48cf76e94c65b1", package.description="all", package.path="opentelemetry-api-1.29.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="a4cf6857f268b9637ea330fffc70c1e6421d1d55", package.description="context", package.path="opentelemetry-context-1.29.0.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="85811b4757731b2fd4d8ff24ec802dcb03b895d5", package.description="instrumentation-api by OpenTelemetry", package.path="opentelemetry-instrumentation-api-1.30.0-SNAPSHOT.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="5e810d79e96c0615813d5ed2859413222c5340b4", package.description="instrumentation-api-semconv by OpenTelemetry", package.path="opentelemetry-instrumentation-api-semconv-1.30.0-alpha-SNAPSHOT.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="8ee51f51d9c1c959b537c8dba67d7a524204b974", package.description="semconv", package.path="opentelemetry-semconv-1.29.0-alpha.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="5cc1be17aed4e1e396c6b5359518f369a42ebc37", package.path="protobuf-java-3.23.4.jar", package.type="jar"}
{event.domain="jvm", event.name="info", package.checksum="6bb59616f1180286bc2ccf40e34d636984581ba9", package.description="metrics", package.path="opentelemetry-sdk-metrics-1.29.0.jar", package.type="jar"}
```

Notice that some jars don't have a version package.name, package.version, or package.description. These fields are missing if we can't find certain things in the JAR, like the manifest "Implementation-Name" and "Implementation-Vendor fields", or the pom.properties file containing the `groupid` and `artifactId`.

We can probably get clever and do a little better than what I have currently, but there are going to be limits.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    if (!seenUris.add(locationUri)) {
+      return;
+    }
+    if ("jrt".equals(jarUrl.getProtocol())) {
+      logger.log(Level.FINEST, "Skipping processing jar for java runtime module: " + jarUrl);
+      return;
+    }
+    if (!jarUrl.getFile().endsWith(JarAnalyzerUtil.JAR_EXTENSION)) {
```
Comment (laurit) [inline instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java:144]: Have you tried to figure out whether there are non-jar urls that could be relevant? I didn't verify but I'd guess that osgi bundles might not point to jar here, war files will probably just point to `WEB-INF/classes` directory on most app servers etc. Could enable this temporarily (or only for smoke tests) and see what urls won't get processed for smoke tests.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    if (!seenUris.add(locationUri)) {
+      return;
+    }
+    if ("jrt".equals(jarUrl.getProtocol())) {
+      LOGGER.log(Level.FINEST, "Skipping processing jar for java runtime module: " + jarUrl);
+      return;
+    }
+    if (!jarUrl.getFile().endsWith(JarAnalyzerUtil.JAR_EXTENSION)) {
+      LOGGER.log(Level.INFO, "Skipping processing jar with unrecognized code location: " + jarUrl);
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java]: No probably not. My thought was we use logging to find all the shortcomings / edge cases with the current process to find / extract metadata. It will be hard to anticipate all the scenarios we'll come across in the wild so just trying to emphasize safety and good logging so we can improve the logic when we encounter issues.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/build.gradle.kts [hunk 4]
```diff
@@ -4,6 +4,14 @@ plugins {
 
 dependencies {
   implementation(project(":instrumentation:runtime-telemetry:runtime-telemetry-java8:library"))
+  implementation("io.opentelemetry:opentelemetry-api-events")
+  implementation("com.google.guava:guava")
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/build.gradle.kts:8]: There's a rate limiter bundled with the SDK: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/common/src/main/java/io/opentelemetry/sdk/internal/RateLimiter.java

Is it safe to use SDK classes in the agent?

If not, do you know of any other places the agent does rate limiting and how they solve the issue? I don't think this needs a sophisticated / precise implementation, but would be good to re-use something if already available.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java (hunk 2 above)
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java:144]: I think its likely that there are other types of locations that warrant trying to handle. Treating logging as a discovery mechanism: 
- if we find other situations which are not useful, we can add logic to intentionally skip them.
- if we find other situations and deem them useful, can add logic to handle them and any challenges.

Comment (laurit) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java:144]: To enable this for smoke tests add it to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/bebd81927771e60954582854ec494f117fd8b600/smoke-tests/src/test/java/io/opentelemetry/smoketest/AbstractTestContainerManager.java#L20-L44

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/build.gradle.kts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/build.gradle.kts]: Went ahead and used the `RateLimiter` from the SDK. Can copy it into instrumentation if we decide we don't want to rely on it.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java (hunk 2 above)
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java]: I enabled this for smoke tests and used the output to refine the processing logic a bit:
- There were code locations that I'm calling "non-archive" based. These are locations that end in `/classes/` or `/jsp/`. I don't know of anyway we can extract / send metadata for these directories, so skipping them with a `FINEST` log message.
- There are `.war` files that I was ignoring. Adjusted the logic to include these.

Here's some filtered output from the smoke tests that show the attributes we manage to extract: https://gist.github.com/jack-berg/d20898c73aa6269d7fd1f766006db90a

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java]: There are no more instances of this message being logged in the smoke tests. What log level do you think is appropriate such that we can discover edge cases with the logic without being noisy for users who don't care?

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+public class JarAnalyzerInstaller implements AgentExtension, AgentListener {
+
+  @Override
+  public void afterAgent(AutoConfiguredOpenTelemetrySdk autoConfiguredOpenTelemetrySdk) {
+    JarAnalyzer.getInstance().maybeInstall(autoConfiguredOpenTelemetrySdk.getOpenTelemetrySdk());
+  }
+
+  @Override
+  @CanIgnoreReturnValue
+  public AgentBuilder extend(AgentBuilder agentBuilder, ConfigProperties config) {
```
Comment (mateuszrzeszutek) [inline instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java:31]: You aren't really using the `agentBuilder` here; is there a need to implement an `AgentExtension`? Wouldn't initializing the jar analyzer in `afterAgent` work?

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    if ("jrt".equals(archiveUrl.getProtocol())) {
+      logger.log(Level.FINEST, "Skipping processing for java runtime module: " + archiveUrl);
+      return;
+    }
+    String file = archiveUrl.getFile();
+    if (file.endsWith("/")) {
+      logger.log(Level.FINEST, "Skipping processing non-archive code location: " + archiveUrl);
+      return;
+    }
+    if (!file.endsWith(JAR_EXTENSION) && !file.endsWith(WAR_EXTENSION)) {
```
Comment (mateuszrzeszutek) [inline instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java:152]: What about `ear`? Some people might still use that

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+   * 30d16ec2aef6d8094c5e2dce1d95034ca8b6cb42}.
+   */
+  static void addPackageChecksum(AttributesBuilder builder, URL archiveUrl) throws IOException {
+    builder.put(PACKAGE_CHECKSUM, computeSha1(archiveUrl));
+  }
+
+  private static String computeSha1(URL archiveUrl) throws IOException {
+    MessageDigest md;
+    try {
+      md = MessageDigest.getInstance("SHA1");
```
Comment (mateuszrzeszutek) [inline instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java:60]: Is `MessageDigest` reusable? Could we perhaps use a single instance of that?

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        JarInputStream jarInputStream = new JarInputStream(inputStream)) {
+      // Advance the jarInputStream to the pom.properties entry
+      for (JarEntry entry = jarInputStream.getNextJarEntry();
+          entry != null;
+          entry = jarInputStream.getNextJarEntry()) {
+        if (entry.getName().startsWith("META-INF/maven")
+            && entry.getName().endsWith("pom.properties")) {
+          if (pom != null) {
+            // we've found multiple pom files. bail!
+            return;
```
Comment (mateuszrzeszutek) [inline instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java:151]: 👍

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java (hunk 6 above)
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java:152]: Oh didn't know about ear. Let me see if I can't get the test module to build an `.ear` so I can confirm the logic works.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java (hunk 7 above)
Comment (laurit) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java:60]: `MessageDigest` has a `reset` method

Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java:60]: Would have to take a lock on it to make sure its safe, but that's not a problem because as of now the processing of jars is single threaded.

Comment (mateuszrzeszutek) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java:60]: A `ThreadLocal` would work fine as well, I think

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java (hunk 5 above)
Comment (mateuszrzeszutek) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java:31]: That makes sense. WDYT about using `BeforeAgentListener` instead? (Or, alternatively, `AgentBuilder#with(TransformerDecorator)` -- but that's probably too much trouble for this quite simple use case)

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java (hunk 6 above)
Comment (laurit) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java:152]: The ear isn't usually an issue because app servers will extract them during the deploy. The exception is jboss that uses a custom url protocol that can represent nested archives. Similarly war files are almost always extracted to support `ServletContext.getRealPath`, I think only weblogic may deploy (there is a checkbox in the weblogic admin console) from packaged war, but that won't affect class loading as they extract WEB-INF/lib and also package classes from WEB-INF/classes into a jar.

**Comment by mateuszrzeszutek** [comment]
It would be nice to add some Spring Boot tests at some point: Spring Boot jars include the dependencies' jar files inside the main uberjar

**Comment by jack-berg** [comment]
> It would be nice to add some Spring Boot tests at some point: Spring Boot jars include the dependencies' jar files inside the main uberjar

Can see the output of this in the smoke tests, right? The output looks normal to be for the spring boot smoke tests.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java (hunk 5 above)
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java:31]: I think `BeforeAgentListener` isn't ideal for the same reason `afterAgent` isn't - by waiting longer in the initialization cycle I think we miss classes.

> Or, alternatively, AgentBuilder#with(TransformerDecorator

What's the advantage of this approach? I'll defer to your expertise on this

Comment (mateuszrzeszutek) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java:31]: > I think `BeforeAgentListener` isn't ideal for the same reason `afterAgent` isn't - by waiting longer in the initialization cycle I think we miss classes.

The `BeforeAgentListener` is called way earlier than the `AgentExtension`; and definitely before the actual agent `ClassFileTransformer` is installed.

> What's the advantage of this approach? I'll defer to your expertise on this

Honestly, I don't know if there's any; maybe being tied into bytebuddy's lifecycle helps in some cases. It's just another option.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+   */
+  static void addPackageDescription(AttributesBuilder builder, URL archiveUrl) throws IOException {
+    try (JarFile jarFile = new JarFile(archiveUrl.getFile())) {
+      Manifest manifest = jarFile.getManifest();
+      if (manifest == null) {
+        return;
+      }
+
+      java.util.jar.Attributes mainAttributes = manifest.getMainAttributes();
+      String name = mainAttributes.getValue(java.util.jar.Attributes.Name.IMPLEMENTATION_TITLE);
```
Comment (tylerbenson) [inline instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java:124]: I noticed you're not getting `Implementation-Version`.  Is that intentional?

Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerUtil.java:124]: The version is being grabbed from the pom, so yes intentional. Although I don't feel particularly strongly about how the `package.description` field is derived.

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java (hunk 5 above)
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzerInstaller.java]: Took this advice and was able to simplify things quite a bit by taking advantage of being able to access Instrumentation and AutoConfiguredOpenTelemetry in the same callback. Nice tip!

File: instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jack-berg) [reply instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java]: I think info is reasonable for now. Feel free to re-open if you think otherwise.

**Comment by jack-berg** [comment]
Hey @open-telemetry/java-instrumentation-maintainers got buried for a couple weeks and lost track of this. Wondering if there's interest in getting this landing (in something close to its current form) and then iterating. There's obviously a lot to do, including starting and seeing through a semantic-conventions discussion, but I think this is a pretty good start and although it will be subject to some churn, can help provide useful insight today.

**Review by tylerbenson — APPROVED** [review]

**Review by mateuszrzeszutek — APPROVED** [review]
